### ISSUE TITLE
native-cpu feature issue with 0.3.0

### ISSUE THREAD

**Issue by zeroflaw** [SchrodingerZhu/snmalloc-rs#169]
I'm unable to build with native-cpu feature, the output contains one failed test, i'm not sure how to output any other useful information.

build works without native-cpu

`  -- Performing Test SNMALLOC_LINKER_SUPPORT_NO_ALLOW_SHLIB_UNDEF
     Called from: [4]   /usr/local/Cellar/cmake/3.23.1_1/share/cmake/Modules/Internal/CheckSourceCompiles.cmake
                  [3]   /usr/local/Cellar/cmake/3.23.1_1/share/cmake/Modules/Internal/CheckLinkerFlag.cmake
                  [2]   /usr/local/Cellar/cmake/3.23.1_1/share/cmake/Modules/CheckLinkerFlag.cmake
                  [1]   /.cargo/git/checkouts/snmalloc-rs-801a4b46a1ccfde5/3eeba56/snmalloc-sys/snmalloc/CMakeLists.txt
  -- Performing Test SNMALLOC_LINKER_SUPPORT_NO_ALLOW_SHLIB_UNDEF - Failed
     Called from: [4]   /usr/local/Cellar/cmake/3.23.1_1/share/cmake/Modules/Internal/CheckSourceCompiles.cmake
                  [3]   /usr/local/Cellar/cmake/3.23.1_1/share/cmake/Modules/Internal/CheckLinkerFlag.cmake
                  [2]   /usr/local/Cellar/cmake/3.23.1_1/share/cmake/Modules/CheckLinkerFlag.cmake
                  [1]  /.cargo/git/checkouts/snmalloc-rs-801a4b46a1ccfde5/3eeba56/snmalloc-sys/snmalloc/CMakeLists.txt`

**Comment by SchrodingerZhu** [comment]
Thanks for the report. Looking into it

**Comment by mjp41** [comment]
@zeroflaw could you provide more details.

* What C++ compiler is it using and its version? clang++-10, g++-9, etc
* What linker are you using? ld.bfd, ld.lld, ld.gold
* Can you provide the full log details?

The message `SNMALLOC_LINKER_SUPPORT_NO_ALLOW_SHLIB_UNDEF` failing is saying that the linker does not support a particular option.  This is not necessarily a problem.  It comes from:
```cmake
  check_linker_flag(CXX "-Wl,--no-undefined" SNMALLOC_LINKER_SUPPORT_NO_ALLOW_SHLIB_UNDEF)
```
This checks that `--no-undefined` can be passed to the linker.  This is used to check that the library is self contained, and mostly ensures that we don't actually need the C++ runtime to run the allocator.

Could you elaborate more on the bad behaviour you are experiencing?

**Comment by mjp41** [comment]
@SchrodingerZhu I have added a PR to check `native-cpu` to the GitHub actions.  #170  I don't have a Mac to test, but I believe this is probably a Mac issue as I think `Cellar` is from `Brew`.

**Comment by zeroflaw** [comment]
I am using a mac
I tried "-C", "link-arg=-fuse-ld=lld" but that didn't help
Can I email the full log?

**Comment by zeroflaw** [comment]
I'm confused its working with:
`snmalloc-rs = { git = "https://github.com/SchrodingerZhu/snmalloc-rs.git", features = ["native-cpu", "build_cc"] }`
fails with:
`snmalloc-rs = { git = "https://github.com/SchrodingerZhu/snmalloc-rs.git", features = ["native-cpu"] }`

**Comment by mjp41** [comment]
`build_cc` doesn't use CMake. The error you are showing is cmake checking if an option is supported or not.

You should be able to attach a log to the message here.  It is better than emailing as everyone can see the details.

**Comment by zeroflaw** [comment]
I found the change that broke it.

Works:
https://github.com/microsoft/snmalloc/blob/6e638742e3c66549174d4c264bd05c9435938ac1/CMakeLists.txt#L196

Fails:
https://github.com/microsoft/snmalloc/blob/d5c732f3c14969d119178cbc8382d592800c5421/CMakeLists.txt#L302


The error is:
`
 target_compile_options called with invalid arguments
`

Not sure why, if you want me to test anything let me know.

**Comment by mjp41** [comment]
Sorry, I was out today.  Could you try changing 
```patch
-  target_compile_options(${name} -march=native)
+  target_compile_options(${name} PRIVATE -march=native)
```
I won't have time to repro and fix locally for a few days.

**Comment by zeroflaw** [comment]
Yes thats seems to work. No problem I'll just skip version 0.3.0 and wait for the next one.
Thanks for your help

**Comment by mjp41** [comment]
This is now fixed upstream.  I have raised a PR to bump the version #171.
